### PR TITLE
Minor update to arch_test.h

### PR DESCRIPTION
**Pull request by yungchinghsiao**
* The TEST_PKRR_OP() macro should read vxsat.OV flag before updating signature
* The _F flag in RDOV() should be updated to zero if Zicsr does not exist

### CODE REVIEW

**Comment by allenjbaum** [comment]
How useful is the P_extension without Zicsr? 
F and D extensions (and others) require Zicsr. Why isn't that the case for P-extension?
If that's the case, some of this isn't necessary.

**Comment by yungchinghsiao** [comment]
> 

In my last PR, @pawks suggested to have a macro to make CSR accesses conditional on Zicsr, citing all tests using csr instructions will have such a test. He might be referring to my access to mstatus in RVTEST_VXSAT_ENABLE() but I did the same treatment for RDOV(). Some of the changes was lost during SIGUPD macro updates so this PR is adding them back.

**Comment by yungchinghsiao** [comment]
> That doesn't answer my question: can anyone usefully implement the Packed extension without Zicsr? If not, then the entire extension should require Zicsr, and these fixes aren't necessary. The spec certainly doesn't mention that vxsat is optional, or that the RDOV or CLROV instructions are optional. Granted, you could have nonconforming ops that perform those functions using an internal state bit - but to be conforming, you'd need a different encoding that does not overlap the CSR op encoding. So: I think you have 3 options: either
> 
> * the overflow flag is optional, or
> * the spec requires Zicsr, or
> * you add new ops with non-CSR encodings to the spec for reading and clearing the OV bit

The overflow flag is only needed by a certain class of DSP algorithms and an implementation can still be useful without the flag for usages that do not require the flag. But as far as the spec is concerned, P should require Zicsr, like V (vxsat is shared by P & V) and it is not very useful to save this bit for an implementation.

The RDOV instruction is defined just as an alias to CSR instructions, so I can remove RDOV()

**Comment by pawks** [comment]
I missed the part where `Zicsr` is a required extension for `P`. In that case the `RVTEST_CASE` macro should capture this dependency and the `RVTEST_ISA` node for all the `P` extension instructions need to be modified to contain the `Zicsr` extension too.

**Comment by yungchinghsiao** [comment]
> I missed the part where `Zicsr` is a required extension for `P`. In that case the `RVTEST_CASE` macro should capture this dependency and the `RVTEST_ISA` macro for all the `P` extension instructions need to be modified to contain the `Zicsr` extension too.

For `RVTEST_ISA` macros: I changed isa string from `IP` to `IPZicsr` for all `P` extension instructions. 

For `RVTEST_CASE` macros, I already have the `Zicsr` checks for all instructions that may set the overflow flag in earlier pull requests. Please see if you are okay with this or shall I include the rest of `P` extension instructions as well.

**Comment by yungchinghsiao** [comment]
> 

Done. I've made all P-ext instructions check for the Zicsr extension.

File: sample_cgfs/rv32ip.cgf [hunk 1]
```diff
@@ -2181,6 +2267,7 @@ zunpkd832:
 # pkbb16:
 #     config:
 #       - check ISA:=regex(.*I.*P.*)
+#       - check ISA:=regex(.*Zicsr.*); def pext_check_vxsat_ov=True
```
Comment (pawks) [inline sample_cgfs/rv32ip.cgf:2270]: Going with the changes in the `template.yaml` shouldn't the check directly check for the `Zicsr` in the previous line itself for all the instructions. The `pext_check_vxsat_ov` macro is no longer relevant right?

Comment (pawks) [reply sample_cgfs/rv32ip.cgf:2270]: Enabling the macro based on `Zicsr` doesn't make sense. Because `Zicsr` is a required extension for `P`. We might need an extra field somewhere in the input isa yaml which indicates whether the optional saturation flag is implemented or not. The macro will have to be enabled based on that field and not `Zicsr`.

**Comment by allenjbaum** [comment]
Shouldn't riscv-config just report that P without Zicsr is an illegal
configuration avoid all this?

File: sample_cgfs/rv32ip.cgf (hunk 1 above)
Comment (yungchinghsiao) [reply sample_cgfs/rv32ip.cgf]: > Enabling the macro based on `Zicsr` doesn't make sense. Because `Zicsr` is a required extension for `P`. We might need an extra field somewhere in the input isa yaml which indicates whether the optional saturation flag is implemented or not. The macro will have to be enabled based on that field and not `Zicsr`. I agree that there needs to be a guard which prevents spurious `mstatus` accesses.

Okay, I removed the pext_check_vxsat_ov check. For the saturation flag, it is not optional in `P`: the spec just does not require mstatus.VS to be implemented, but requires it to guard against vxsat accesses if implemented. Since VS is WARL, there is no need to have extra checks against mstatus.

**Comment by allenjbaum** [comment]
So if VS is implemented, then a trap occurs if vxsat is accessed and VS is
inOFF state
     if VS is implemented  then                          vxsat is accessed
if VS isn't in OFF state,
     if VS is not implemented then a trap occurs if vxsat is accessed.
Correct?
  I'm a bit surprised that VS is WARL, since FS is a WARL field, yet the
priv spec says about FS: (paraphrased)

If the F extension is implemented, the FS field shall not be hardwired to
zero
else if S-mode is implemented, then   FS is hardwired to zero.
else                                                      FS may optionally
be hardwired to zero.

Is there any reason the behavior should be different?

**Comment by pawks** [comment]
> Shouldn't riscv-config just report that P without Zicsr is an illegal
> configuration avoid all this?
Yes, riscv-config will catch the dependency between Zicsr and P. Having the check in the tests also is more of a failsafe rather than a necessity now.


Looks like the behavior of the VS flag is similar to that of the FS flag, the only difference being that VS is optional for P extension and mandatory for V extension. Since the `Zicsr` is a necessary extension, we can include the setting of the VS field in the `mstatus` unconditionally in the P tests. If the field is not implemented, the ideal behavior would be `unchanged`. However, I believe that spurious accesses should be avoided, especially because it is a WARL field and some edge case might break the tests. The ideal way forward is to define `VS` field in riscv-config similar to the `FS` field (defined [here](https://github.com/riscv-software-src/riscv-config/blob/138c69e1716b085702f8984453b203ce8e36a68c/riscv_config/schemas/schema_isa.yaml#L672)) and then conditionally enable the mstatus operation based on whether the `implemented` field in that node is set to `True`. This will involve adding the filed in riscv-config and an API for testing whether the field is implemented in riscof. So the config node ends up looking like the following:
```
- check ISA:=regex(.*I.*P.*Zicsr.*) # Zicsr can be omitted from here.
- check is_implemented(mstatus::VS,$xlen):=True; def pext_check_vxsat_ov=True
```
An alternative can also be:
```
- check ISA:=regex(.*I.*P.*Zicsr.*); def pext_check_vxsat_ov=is_implemented(mstatus::VS,$xlen) # Zicsr can be omitted from here.
```

**Comment by yungchinghsiao** [comment]
@allenjbaum @pawks I found that the order of the `offset` adjustment code in *SIGUPD are updated in riscv-arch-test  [(issue 240)](https://github.com/riscv-non-isa/riscv-arch-test/issues/240). Should the same change be ported over to riscv-ctg?

**Comment by pawks** [comment]
Yes. It needs to be ported over. Once the PR is merged onto the master, the changes will be copied here too.

**Comment by allenjbaum** [comment]
I certainly didn't intend to change any ordering, for what it is worth. I
thought my changes kept them in the same order, though it wouldn't be the
first time I got something wrong :-)

**Comment by yungchinghsiao** [comment]
> > Shouldn't riscv-config just report that P without Zicsr is an illegal
> > configuration avoid all this?
> > Yes, riscv-config will catch the dependency between Zicsr and P. Having the check in the tests also is more of a failsafe rather than a necessity now.
> 
> Looks like the behavior of the VS flag is similar to that of the FS flag, the only difference being that VS is optional for P extension and mandatory for V extension. Since the `Zicsr` is a necessary extension, we can include the setting of the VS field in the `mstatus` unconditionally in the P tests. If the field is not implemented, the ideal behavior would be `unchanged`. However, I believe that spurious accesses should be avoided, especially because it is a WARL field and some edge case might break the tests. The ideal way forward is to define `VS` field in riscv-config similar to the `FS` field (defined [here](https://github.com/riscv-software-src/riscv-config/blob/138c69e1716b085702f8984453b203ce8e36a68c/riscv_config/schemas/schema_isa.yaml#L672)) and then conditionally enable the mstatus operation based on whether the `implemented` field in that node is set to `True`. This will involve adding the filed in riscv-config and an API for testing whether the field is implemented in riscof. So the config node ends up looking like the following:
> 
> ```
> - check ISA:=regex(.*I.*P.*Zicsr.*) # Zicsr can be omitted from here.
> - check is_implemented(mstatus::VS,$xlen):=True; def pext_check_vxsat_ov=True
> ```
> 
> An alternative can also be:
> 
> ```
> - check ISA:=regex(.*I.*P.*Zicsr.*); def pext_check_vxsat_ov=is_implemented(mstatus::VS,$xlen) # Zicsr can be omitted from here.
> ```

@pawks I tried your suggestions and made modifications to riscv-config. However, it seems riscof do not understand is_implemented() and it won't define  pext_check_vxsat_ov. From what I could see in riscof code, the function is not support, is this expected?

**Comment by pawks** [comment]
Yes. The `is_implemented` function is not implemented in riscof yet. That was just an example of how we can achieve this. The set of functions necessary to write the `check` conditions in `RVTEST_CASE` macro is a WIP and not complete. The `is_implemented` function will have to be added [here](https://github.com/riscv-software-src/riscof/blob/e78f8dac0abbc40d2b8b7a66a8db9bb6eaf3dd94/riscof/framework/test.py#L83).

**Comment by yungchinghsiao** [comment]
> Yes. The `is_implemented` function is not implemented in riscof yet. That was just an example of how we can achieve this. The set of functions necessary to write the `check` conditions in `RVTEST_CASE` macro is a WIP and not complete. The `is_implemented` function will have to be added [here](https://github.com/riscv-software-src/riscof/blob/e78f8dac0abbc40d2b8b7a66a8db9bb6eaf3dd94/riscof/framework/test.py#L83).

I see. So is there anything I should do to get this pull request accepted?

**Comment by allenjbaum** [comment]
Just to be clear: what is the justification of making VS optional? It isn't optional for F/d/H extensions; why should it be optional here?

**Comment by yungchinghsiao** [comment]
> Just to be clear: what is the justification of making VS optional? It isn't optional for F/d/H extensions; why should it be optional here?

You should probably ask this question in the P workgroup. My understanding was there is only one state bit in P. There were no requirements for state management but it was suggested to share the vxsat.OV with `V` for this state bit. So the clause was written in this way to hint that the `VS` behavior is only needed if `V` is implemented.

**Comment by allenjbaum** [comment]
I guess the real question is "why is V bit optional" then

**Comment by Kevin-Andes** [comment]
Dear Allen, 

Per our discussion on 04/12/2022 (right after Fast Interrupt TG meeting), `mstatus.VS` flag is optional for P-ext because it won't benefit P-ext at all. This is because P-ext only uses 1 CSR (i.e., `vxsat`) so it is much faster to simply always save/restore this 1 CSR in context-switch rather than looking up `mstatus.VS` first (which requires more efforts: read CSR, extract the VS field, check and branch on its value). Therefore, `mstatus.VS` flag is optional for P-ext and this was previously reviewed/decided by the Architecture Review Committee.

Also, the `vxsat` CSR is mandatory for P-ext in case it isn't clear from the spec.

Please feel free to let us know if you have any more questions or concerns. Thanks.

   Best regards,
   Kevin

> I guess the real question is "why is V bit optional" then

**Comment by pawks** [comment]
> > Yes. The `is_implemented` function is not implemented in riscof yet. That was just an example of how we can achieve this. The set of functions necessary to write the `check` conditions in `RVTEST_CASE` macro is a WIP and not complete. The `is_implemented` function will have to be added [here](https://github.com/riscv-software-src/riscof/blob/e78f8dac0abbc40d2b8b7a66a8db9bb6eaf3dd94/riscof/framework/test.py#L83).
> 
> I see. So is there anything I should do to get this pull request accepted?

I guess this PR can be merged then. The caveat is that the `pext_check_vxsat_ov` flag will never be set until `riscof` gets support for the `is_implemented` function. The tests themselves need not change further. IMO, this PR can be merged.

**Review by pawks — APPROVED** [review]

**Comment by allenjbaum** [comment]
Let's ask a more basic question: why should VS ever be optional?
What is the cost of implementing it, and the cost of not implementing it?
The reason it exists at all is to greatly lower the cost of context
switches by the OS (or hypervisor)
The cost of not implementing it is much more context switch overhead
proportional to many registers affect these bits.
(so if any write to an XREG can affect these the cost is 32x if it affects
a single CSR)
What is the cost of implementing it? PRobably a handful of gates and 2
flops.

**Comment by yungchinghsiao** [comment]
Hello Allen,

Your question is a good one but it should probably be addressed to the
P-ext task group.

Yung-ching
